### ISSUE TITLE
about AnoDDPM

### ISSUE THREAD

**Issue by tommyq666** [Taurusmumu/AnoPILaD#1]
Dear author,

I noticed that you mentioned the work AnoDDPM (Linmans, Jasper, et al. "Diffusion models for out-of-distribution detection in digital pathology." Medical Image Analysis 93 (2024): 103088) in your paper and compared your method with it. However, since this paper has not released its code, I’m unable to reproduce the results. Could you please let me know how you reproduced this part of the work, or if possible, share the corresponding code? Thank you very much!

**Comment by Taurusmumu** [comment]
@tommyq666 Hi, I will upload my AnoDDPM implementation! Probably by this week.

**Comment by tommyq666** [comment]
Thank you very much!

**Comment by Taurusmumu** [comment]
@tommyq666 Hi, I just uploaded my AnoDDPM implementation! Training and image generation are basically the same as the procedure that is written in AnoPILaD ReadMe. Please run the scripts before you modify the config file.

**Comment by tommyq666** [comment]
thanks!

**Comment by tommyq666** [comment]
Hello, may I ask — in the training command:

```
CUDA_VISIBLE_DEVICES=0,1 OMP_NUM_THREADS=2 python -m torch.distributed.run --nproc_per_node=2 train_text_to_image_lora.py --train_batch_size 16 --num_train_epochs 20 --train_data_dir "YOUR_TRAIN_DATA_DIR" --output_dir "./pretrained_weights/sd-model-finetuned-lora/" --validation_prompt ['prompt0', 'prompt1',] --num_validation_images 4
```

What should be written in `--validation_prompt ['prompt0', 'prompt1',]`?
Are there any specific requirements for these prompts?

**Comment by Taurusmumu** [comment]
@tommyq666 To acquire the prompts for images, the Image Captioning step is supposed to be executed first. Through the Image Captioning step, you can see what prompts look like, e.g., "a histopathology image of lymphocyte".

**Comment by tommyq666** [comment]
<img width="631" height="355" alt="Image" src="https://github.com/user-attachments/assets/1e92e683-a7db-4c19-ae24-8416edec027e" />

I’ve run the Image Captioning model and obtained image captions for my training set. However, since there are many images and corresponding captions, should I just select a few captions at random?

**Comment by Taurusmumu** [comment]
@tommyq666 In my case, I just chose several frequent prompts for the validation to check the training quality.

**Comment by tommyq666** [comment]
Hello, I have a few questions regarding the code you provided:

**Question 1: Selection of validation prompts**
During training, I used the following command:

```bash
CUDA_VISIBLE_DEVICES=1 python train_text_to_image_lora.py \
  --pretrained_model_name_or_path "./CompVis-stable-diffusion-v1-4_pretrained_stable_diffusion/2880f2ca379f41b0226444936bb7a6766a227587" \
  --train_batch_size 16 \
  --num_train_epochs 20 \
  --train_data_dir "/home/qiuhongyi/AnoPILaD-main/data/train" \
  --output_dir "./pretrained_weights/sd-model-finetuned-lora/" \
  --validation_prompt "['a histopathology image of lymphocyte', 'a histopathology image of small dormant lymphocytes', 'a histopathology image of follicular dendritic cells', 'a histopathology image of prominent single nucleolus']" \
  --num_validation_images 4
```

Is this selection of `validation_prompt` appropriate, or should I include more prompts?
Also, I am a bit confused: when using prompts to guide reconstruction, does the model try to generate an image containing all the features from the prompts simultaneously? If so, could this lead to discrepancies from the original image? For example, if the input image is **“a histopathology image of lymphocyte”**, the reconstructed image might instead show features of **“a histopathology image of small dormant lymphocytes”**, which would not fully match the original.

After training, I obtained the model weights:
`./AnoPILaD-main/pretrained_weights/sd-model-finetuned-lora/pytorch_lora_weights.safetensors`
However, during validation, I saw the following message:

```
Generating 4 images with prompt: ['[a histopathology image of lymphocyte,', 'a histopathology image of small dormant lymphocytes,', 'a histopathology image of follicular dendritic cells,', 'a histopathology image of prominent single nucleolus]'].
Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
```

<img width="598" height="103" alt="Image" src="https://github.com/user-attachments/assets/3decd36d-b912-4f72-aa0f-543eb9a7df9c" />

Is this normal?

**Question 2: Purpose of the validation dataset**
The README mentions four data folders: YOUR_TRAIN_DIR, YOUR_VAL_DIR, YOUR_TEST_ID_DIR and YOUR_TEST_OOD_DIR. However, it seems that only the training set is used during training. What is the specific role of the validation dataset?

**Question 3: Evaluation metrics and anomaly heatmaps**
I ran the following command to reconstruct images:

```bash
CUDA_VISIBLE_DEVICES=0 python rec_generate.py
```

I generated reconstructed images for the validation, test_ID, and test_OOD datasets, but I did not obtain AUROC or other evaluation metrics, nor the corresponding anomaly heatmaps. Could you please provide the **patch-level** and **WSI-level** evaluation code?

Thank you very much for your guidance!

**Comment by tommyq666** [comment]
Dear Author,
I hope this message finds you well. I've been working with your "AnoDDPM" code this week and truly appreciate you making it available.
However, I noticed that the repository doesn't seem to include the evaluation code for calculating AUROC at both the patch level and the WSI level. Would it be possible for you to provide this? It would be immensely helpful for my research.
Thank you for your time and consideration. Any assistance you could offer would be greatly appreciated！

**Comment by Taurusmumu** [comment]
@tommyq666 

As for question 1:
"
Also, I am a bit confused: when using prompts to guide reconstruction, does the model try to generate an image containing all the features from the prompts simultaneously? If so, could this lead to discrepancies from the original image? For example, if the input image is “a histopathology image of lymphocyte”, the reconstructed image might instead show features of “a histopathology image of small dormant lymphocytes”, which would not fully match the original.
"
If I understand your question correctly, you concerned that when a batch of prompts is provided, the outputs are affected by all the input prompts. This will not happen; for i in range(4), the output[i] is affected only by prompt[i]. 

To solve your screenshot, you can disable the safety checker by doing:
"
pipe = DiffusionPipeline.from_pretrained("google/ddpm-celebahq-256",
                                             unet=unet,
                                             safety_checker=None,
                                             requires_safety_checker=False)
"

Question 2:
The validation dataset is used to calculate the z-score.

Question 3:
I will upload my evaluation code!

**Comment by tommyq666** [comment]
thank you !

**Comment by tommyq666** [comment]
Dear Author,I apologize for reaching out again.I am highly interested in the evaluation methods presented in your paper.Could you please let me know the estimated release date for the corresponding source code used in your evaluation?Thank you very much for your time and consideration. I look forward to your reply！

**Comment by tommyq666** [comment]
I have another question regarding the C16 test set you used. For the abnormal patches, are they entirely composed of tumor regions, or could an abnormal patch still contain only a partial tumor region?

In addition, when evaluating at the WSI (Whole Slide Image) level, should the entire WSI be divided into 256×256 patches and then arranged according to their spatial order to compute anomaly scores for each patch, followed by aggregating them back to the WSI to generate a heatmap? If aggregation is required, could you please clarify the specific aggregation method?
